### PR TITLE
Update to support lightning (new device API) & the PL operator arithmetic update

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
- PL updated the operator arithmetic
- PL-Lightning switched to the new device API

**Description of the Change:**
- Update the transform pipeline for the new device API
- Create a decompose compatible with Catalyst
- Update for the operator arithmetic
- Pin PL to master and PL-lightning to dev-30

**Benefits:**
- Catalyst supports the updated version of PL and Ligthning

### CODE REVIEW

**Comment by rmoyard** [comment]
Changes needed: https://github.com/PennyLaneAI/pennylane/pull/5506 https://github.com/PennyLaneAI/pennylane/pull/5478

File: frontend/test/lit/test_measurements.py
Comment (rmoyard) [inline frontend/test/lit/test_measurements.py:176]: Revert after https://github.com/PennyLaneAI/pennylane/pull/5478

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [inline frontend/catalyst/jax_tracer.py:519]: Op arithmetic update

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [inline frontend/catalyst/jax_tracer.py:531]: Move pauli rep at the end because op that were previously hamiltonian are now sum

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [inline frontend/catalyst/jax_tracer.py:623]: Wrong wire number

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [inline frontend/catalyst/preprocess.py:69]: Very similar to PL decompose

File: frontend/catalyst/qjit_device.py
Comment (rmoyard) [inline frontend/catalyst/qjit_device.py:294]: Merge decompose unitary with decompose in a single function.

File: frontend/test/lit/test_custom_ops.py
Comment (rmoyard) [inline frontend/test/lit/test_custom_ops.py:40]: It just uses kokkos list of operations because the one from lightning is a frozen set.

File: frontend/test/lit/test_decomposition.py
Comment (rmoyard) [inline frontend/test/lit/test_decomposition.py:30]: It just uses kokkos list of operations because the one from lightning is a frozen set.

File: frontend/test/lit/test_quantum_control.py
Comment (rmoyard) [inline frontend/test/lit/test_quantum_control.py:33]: It just uses kokkos list of operations because the one from lightning is a frozen set.

File: frontend/test/pytest/test_decomposition.py
Comment (rmoyard) [inline frontend/test/pytest/test_decomposition.py:25]: It just uses kokkos list of operations because the one from lightning is a frozen set.

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [inline frontend/catalyst/preprocess.py:100]: @lillian542 Notice here I add to add the right Jax fram otherwise we have leaked values issues.

**Comment by rmoyard** [comment]
[sc-60562]

**Comment by rmoyard** [comment]
[sc-60606]

**Comment by rmoyard** [comment]
[sc-60561]

File: .dep-versions
Comment (maliasadi) [inline .dep-versions:9]: Could you please update this to dev32 to include some of the decomposition and serialization fixes for LQ and LK?
 
```suggestion
lightning=0.36.0-dev32
```

File: doc/requirements.txt
Comment (maliasadi) [inline doc/requirements.txt:32]:
```suggestion
pennylane-lightning==0.36.0-dev32
```

File: frontend/catalyst/jax_tracer.py
Comment (maliasadi) [inline frontend/catalyst/jax_tracer.py:277]:
```suggestion
```

File: frontend/test/lit/test_custom_ops.py
Comment (maliasadi) [reply frontend/test/lit/test_custom_ops.py:40]: Catalyst integration tests should be rely on LQ by default so that users can run these tests without having LK installed on their machine. Can we cast the frozenset of LQ ops to a list here?

**Review by maliasadi — COMMENTED** [review]
Thanks @rmoyard! :partying_face:

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [reply frontend/catalyst/jax_tracer.py:277]: This one is important for the decomposition

File: frontend/test/lit/test_custom_ops.py
Comment (rmoyard) [reply frontend/test/lit/test_custom_ops.py:40]: Good point let me change that

Comment (rmoyard) [reply frontend/test/lit/test_custom_ops.py]: Updated everywhere

File: frontend/catalyst/jax_tracer.py
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py:277]: could you add a brief code comment here for the future reference?

Comment (rmoyard) [reply frontend/catalyst/jax_tracer.py:277]: Yes for sure!

**Review by maliasadi — APPROVED** [review]
All good from my side :100: you may want to wait for the second approval from the compilation team.

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [reply frontend/catalyst/jax_tracer.py:278]: Done

Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py:278]: Thanks!

File: frontend/catalyst/preprocess.py
Comment (lillian542) [reply frontend/catalyst/preprocess.py:100]: Where does the jax frame come from when calling `decompose`?

**Comment by erick-xanadu** [comment]
@rmoyard the shortcut ticket here says that this issue is related to decomposing nested operations like control and adjoint, but there are no tests that check for the decomposition of them. Is there an error in the shortcut ticket assigned to this?

**Comment by rmoyard** [comment]
@erick-xanadu They are multiple shortcut tickets (this is the part one for the decompose ticket), the unit tests for decompose will be added by @lillian542 as the part 2, she is currently working on making the solution better.

File: frontend/catalyst/preprocess.py
Comment (erick-xanadu) [reply frontend/catalyst/preprocess.py:70]: Have you considered outlining some parts of this method to make it more readable. For example:

```python
def decompose_to_matrix(op):
   if !op.name in {"MultiControlledX", "BlockEncode"} ...
       return None
   ... # Continue with normal path
   
def decompose_hybrid_op(op):
   if !isinstance(op, HybridOp):
       return None
   ... # Continue with normal path
```

and then

```python
def decomposer(op):
   success = decompose_to_matrix(op)
   if success:
       return success
   success = decompose_hybrid_op(op)
   if success:
       return success
   return op.decomposition()
```

Also, is there a reason why the decomposer needs to be a nested function? Are you making a closure out of some arguments?

File: frontend/catalyst/preprocess.py
Comment (erick-xanadu) [inline frontend/catalyst/preprocess.py:97]:
```suggestion
    elif isinstance(op, catalyst.jax_tracer.HybridOp):
```

I think the above should work.

File: frontend/catalyst/preprocess.py
Comment (erick-xanadu) [inline frontend/catalyst/preprocess.py:180]:
```suggestion
    elif isinstance(op, catalyst.jax_tracer.HybridOp):
```

I think the above should work.

File: frontend/test/lit/test_quantum_control.py
Comment (erick-xanadu) [inline frontend/test/lit/test_quantum_control.py:69]:
```suggestion
            return "DummyDevice", get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/libdummy_device.so"
```

I think this is not being tested. The `get_c_interface` method changed APIs a while back and instead of returning just the path, it should return the name and the path.

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [reply frontend/catalyst/preprocess.py:180]: No it does not unfortunately because QCtrl handle decomposition differently by defining `.decomposition()`

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [reply frontend/catalyst/preprocess.py:70]: Ok no worries I do that. Because decomposition potentially need a big depth, template to some high level op to a universal gate set. You check the decompose in PL it is very similar.

File: frontend/test/lit/test_quantum_control.py
Comment (rmoyard) [reply frontend/test/lit/test_quantum_control.py]: Done

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [reply frontend/catalyst/preprocess.py:70]: Changes done!

File: frontend/catalyst/preprocess.py
Comment (erick-xanadu) [reply frontend/catalyst/preprocess.py:90]: Would it be possible to add a decomposition method to these? Also, does decomposing QCtrl be visited multiple times or that is not the case?

Comment (rmoyard) [reply frontend/catalyst/preprocess.py:90]: The decomposition for hybrid op is just to visit the tape of each region

Comment (erick-xanadu) [reply frontend/catalyst/preprocess.py:90]: Yes, but I think (maybe if the ctx was stored in the op itself) we could have a:

```python
class HybridOp:

    def decomposition(self):
        return _decompose_hybrid_op(self, self.ctx, self.stopping_condition, self.max_expansion)
```

Or something similar.

File: frontend/catalyst/jax_primitives.py
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:868]: If we only generate ranked tensors, how about instead of either of the proposals, we do something like the following:

```python
if not rankedTensor:
  raise TypeError("a ranked tensor is expected")

... # the normal path
```

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [reply frontend/catalyst/preprocess.py:90]: The thing is that the hybrid Op has no ctx attached to it, it is something that can be passed from the tracing function. So the decomposition method would have a different signature and then you  to handle the case differently in the decomposer.

File: frontend/catalyst/jax_primitives.py
Comment (rmoyard) [reply frontend/catalyst/jax_primitives.py:868]: My change just makes sure that we never have undefined `baseType` variable, I will remove the change and we can handle it in a different PR.

File: frontend/catalyst/jax_tracer.py
Comment (rmoyard) [reply frontend/catalyst/jax_tracer.py:278]: For each op, it checks if it is accepted (in the list of operations or an hybrid op that is visited), if it is not the case it decomposes recursively until it either have all accepted operations or fails if it is reaching the max expansion. The hybrid operation is handled slightly differently, the decomposition is to call `decompose` on tapes for each region, if it can decompose all of them then it succeed and we mark the operation as visited, which tells that we successfully expanded the hybrid op. Is it more clear?

File: frontend/catalyst/preprocess.py
Comment (rmoyard) [reply frontend/catalyst/preprocess.py]: There is a trace associated to each region for the hybrid op, `with EvaluationContext.frame_tracing_context(ctx, region.trace):` gives the right scope for the decomposition to happen. Otherwise it considers that the tracer are leaked when in the decomposition you actually uses the parameters: e.g. `op -> [RX(phi / 8, wires=0), RX(phi / 8, wires=0)]`

File: frontend/catalyst/preprocess.py
Comment (erick-xanadu) [reply frontend/catalyst/preprocess.py:90]: Do you think changing the definition of HybridOp to store the context is not possible to affect the decomposition?

I know at the very least all HybridOps have this method:

```
class ForLoop(HybridOp):

    def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
```

which if it is called before decomposition, then we can store the context in self.

Comment (rmoyard) [reply frontend/catalyst/preprocess.py:96]: Trace quantum is called after the transforms, so the only way is to change the signature of `decomposition(...` and switch in the decomposer which does not make things cleaner. Also @lillian542 has a ticket to make the implementation more robust https://app.shortcut.com/xanaduai/story/61450/catalyst-can-decompose-nested-operations-hybrid-op-part-2

**Review by erick-xanadu — APPROVED** [review]
Thanks Romain!

File: frontend/catalyst/preprocess.py
Comment (dime10) [reply frontend/catalyst/preprocess.py:96]: @erick-xanadu For loop ops are not meant to be decomposed, and a decomposition yielding another For loop op would lead to an infinite recursion.

File: frontend/catalyst/preprocess.py
Comment (dime10) [reply frontend/catalyst/preprocess.py:194]: I think what we should do here is have a standard property where nested tapes are stored. Any op with that property can then be recursed into (assuming it wasn't decomposed first).

**Pull request merged**
